swap/swap.go

```diff
@@ -17,69 +17,340 @@
 package swap
 
 import (
+	"context"
+	"crypto/ecdsa"
 	"errors"
 	"fmt"
+	"math/big"
 	"strconv"
 	"sync"
+	"time"
 
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/crypto"
 	"github.com/ethereum/go-ethereum/p2p/enode"
+	"github.com/ethersphere/swarm/contracts/swap"
+	cswap "github.com/ethersphere/swarm/contracts/swap"
 	"github.com/ethersphere/swarm/log"
 	"github.com/ethersphere/swarm/p2p/protocols"
 	"github.com/ethersphere/swarm/state"
 )
 
-// SwAP Swarm Accounting Protocol
+// ErrInvalidChequeSignature indicates the signature on the cheque was invalid
+var ErrInvalidChequeSignature = errors.New("invalid cheque signature")
+
+// Swap represents the SwAP Swarm Accounting Protocol
 // a peer to peer micropayment system
 // A node maintains an individual balance with every peer
 // Only messages which have a price will be accounted for
 type Swap struct {
-	stateStore state.Store        //stateStore is needed in order to keep balances across sessions
-	lock       sync.RWMutex       //lock the balances
-	balances   map[enode.ID]int64 //map of balances for each peer
+	api                 PublicAPI
+	stateStore          state.Store          // stateStore is needed in order to keep balances across sessions
+	lock                sync.RWMutex         // lock the balances
+	balances            map[enode.ID]int64   // map of balances for each peer
+	cheques             map[enode.ID]*Cheque // map of balances for each peer
+	peers               map[enode.ID]*Peer   // map of all swap Peers
+	backend             cswap.Backend        // the backend (blockchain) used
+	owner               *Owner               // contract access
```

Comment:
does this need to be a pointer?

Reply:
Well we pass it around as part of the swap instance quite a bit, why *not* a pointer?

Reply:
because a pointer infers you would potentially want to modify the content of the value you're pointing to. is there a case where the `Owner` of the `Swap` instance should change? or any of the details of the `Owner` struct should be changed on the fly?

Reply:
As an example, when created, `owner.Contract` may be empty, it will only be assigned when the contract is deployed. But in a next iteration we will most probably allow to pass a contract address via configuration, so that it can be assigned an existing contract.

We also have to assign values to `owner` a lot for tests, you may want to check `swap_test.go` for references.